bpo-39184: Add audit events to functions in `fcntl`, `msvcrt`, `os`, `resource`,  `shutil`, `signal`, `syslog`

gousaiyang opened the pull request:
Add audit events to the following functions:

- all functions in `fcntl` and `syslog`
- file operations in `msvcrt`, `os` and `shutil`
- os.putenv/unsetenv
- os.add_dll_directory
- os.fork/forkpty
- os.kill/killpg
- resource.setrlimit, resource.prlimit
- signal.pthread_kill

<!-- issue-number: [bpo-39184](https://bugs.python.org/issue39184) -->
https://bugs.python.org/issue39184
<!-- /issue-number -->

gousaiyang committed:
bpo-39184: Add audit events to functions in `fcntl`, `msvcrt`, `os`, `resource`,  `shutil`, `signal`, `syslog`

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gousaiyang committed:
Use explicit 1-tuple for PySys_Audit

zooba commented on Modules/fcntlmodule.c:
```diff
@@ -66,6 +66,11 @@ fcntl_fcntl_impl(PyObject *module, int fd, int code, PyObject *arg)
     int async_err = 0;
 
+    if (PySys_Audit("fcntl.fcntl", "iiO", fd, code,
+                    arg ? arg : PyLong_FromLong(0)) < 0) {
```
```suggestion
                    arg ? arg : Py_None) < 0) {
```

Otherwise we will cause a reference leak - the `PySys_Audit` call will do its own incref/decref on the arguments, which will leave our PyLong_FromLong reference behind on return.

zooba commented on Modules/fcntlmodule.c:
```diff
@@ -171,6 +176,11 @@ fcntl_ioctl_impl(PyObject *module, int fd, unsigned int code,
     char buf[IOCTL_BUFSZ+1];  /* argument plus NUL byte */
 
+    if (PySys_Audit("fcntl.ioctl", "iIO", fd, code,
+                    ob_arg ? ob_arg : PyLong_FromLong(0)) < 0) {
```
```suggestion
                    ob_arg ? ob_arg : Py_None) < 0) {
```

zooba commented on Modules/fcntlmodule.c:
```diff
@@ -372,6 +386,13 @@ fcntl_lockf_impl(PyObject *module, int fd, int code, PyObject *lenobj,
     int async_err = 0;
 
+    if (PySys_Audit("fcntl.lockf", "iiOOi", fd, code,
+                    lenobj ? lenobj : PyLong_FromLong(0),
```
```suggestion
                    lenobj ? lenobj : Py_None,
```

zooba commented on Modules/fcntlmodule.c:
```diff
@@ -372,6 +386,13 @@ fcntl_lockf_impl(PyObject *module, int fd, int code, PyObject *lenobj,
 
+    if (PySys_Audit("fcntl.lockf", "iiOOi", fd, code,
+                    lenobj ? lenobj : PyLong_FromLong(0),
+                    startobj ? startobj : PyLong_FromLong(0),
```
```suggestion
                    startobj ? startobj : Py_None,
```

zooba commented on Modules/posixmodule.c:
```diff
@@ -3007,6 +3015,13 @@ os_chmod_impl(PyObject *module, path_t *path, int mode, int dir_fd,
 
+    if (PySys_Audit("os.chmod", "OiO",
+                    path->object ? path->object : Py_None, mode,
+                    dir_fd == DEFAULT_DIR_FD ? Py_None : PyLong_FromLong(dir_fd)
```
We may just need to pick an "obvious" value to pass rather than None here... it's a shame that DEFAULT_DIR_FD is so arbitrary.

I'd rather not have to manage an object's lifetime around the PySys_Audit call. Passing `0` or `-1` (assuming those aren't valid FDs? Are they?) would be obvious enough to me.

gousaiyang replied:
According to documentation from [Linux](http://man7.org/linux/man-pages/man2/open.2.html), [FreeBSD](https://www.freebsd.org/cgi/man.cgi?query=open&sektion=2), [macOS](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/open.2.html) and [Microsoft C Run-Time Library](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/open-osfhandle?view=vs-2019), file descriptors are non-negative integers, negative values are used to indicate errors, so here we can safely use `-1` to indicate that `dir_fd` is not specified.

zooba commented on Modules/posixmodule.c:
```diff
@@ -2911,6 +2911,11 @@ os_chdir_impl(PyObject *module, path_t *path)
     int result;
 
+    if (PySys_Audit("os.chdir", "(O)",
+                    path->object ? path->object : Py_None) < 0) {
```
Pretty sure `path->object` can't be NULL, right? Have you seen a case where this is not true?

I mean, there isn't another way to provide the path if you don't pass an object in...

zooba replied:
I see there are a few existing cases where we check it like this, but AFAICT none of them should be necessary.

gousaiyang replied:
For path-like arguments with default values (like `os.listdir` and `os.scandir`), checking `path->object` for `NULL` is necessary, otherwise `SystemError: NULL object passed to Py_BuildValue` could happen. The audit hooks of [`os.listdir`](https://github.com/python/cpython/commit/60419a7e96577cf783b3b45bf3984f9fb0d7ddff#diff-a6f29e907cbb5fffd44d453bcd7b77d5R3763) and [`os.scandir`](https://github.com/python/cpython/commit/60419a7e96577cf783b3b45bf3984f9fb0d7ddff#diff-a6f29e907cbb5fffd44d453bcd7b77d5R13172) are already doing this.
For path-like arguments without default values, it seems that `path->object` could never be `NULL` since the argument is required. The [audit hook in `os.open`](https://github.com/python/cpython/commit/b82e17e626f7b1cd98aada0b1ebb65cb9f8fb184#diff-a6f29e907cbb5fffd44d453bcd7b77d5R8292) does not check `path->object` for `NULL`, and the `os_readlink_impl` function actually assumed that `path->object` is not `NULL` (otherwise the `PyUnicode_Check` call will crash by dereferencing a null pointer).

zooba replied:
Okay, that sounds good to me. Thanks!

zooba reviewed (changes requested):
Overall, looks great! I checked all the events against the doc updates, and they all seem to be in good spots in the code.

We definitely should fix the `PyLong_FromLong` calls to avoid reference leaks.

I'm also 99% sure that we don't need to check `path->object` for `NULL`, but if you come back and tell me that we need to then I'll believe you :)

gousaiyang committed:
Review on PR 18407
- Do not use `PyLong_FromLong`
- Use -1 when `dir_fd` is not specified
- Remove unnecessary NULL checks
